src/web/app/pages-student/student-home-page/student-home-page.component.ts

```diff
@@ -94,67 +96,81 @@ export class StudentHomePageComponent implements OnInit {
-    this.courseService.getAllCoursesAsStudent().subscribe((resp: Courses) => {
-      if (!resp.courses.length) {
-        this.isCoursesLoading = false;
-      }
-      for (const course of resp.courses) {
-        this.feedbackSessionsService.getFeedbackSessionsForStudent('student', course.courseId)
-          .subscribe((fss: FeedbackSessions) => {
-            const sortedFss: FeedbackSession[] = this.sortFeedbackSessions(fss);
+    this.courseService.getAllCoursesAsStudent()
+      .pipe(finalize(() => { this.isCoursesLoading = false; }))
+      .subscribe((resp: Courses) => {
+        for (const course of resp.courses) {
+          this.courses.push({
+            course,
+            feedbackSessions: [],
+            isFeedbackSessionsLoading: true,
+            hasFeedbackSessionsLoadingFailed: false,
+          });
+          this.courses.sort((a: StudentCourse, b: StudentCourse) =>
+              ((a.course.courseId > b.course.courseId) ? 1 : -1));
+          this.loadFeedbackSessionsForCourse(course.courseId);
+        }
+      }, (e: ErrorMessageOutput) => {
+        this.hasCoursesLoadingFailed = true;
+        this.statusMessageService.showErrorToast(e.error.message);
+      });
+    }
+
+  /**
+   * Load feedback sessions for a single course.
+   * The course should have been pushed to the this.courses array before this.
+   */
+  loadFeedbackSessionsForCourse(courseId: string): void {
+    // reference to the course within the this.courses array
+    const courseRef = this.courses.find((c) => c.course.courseId === courseId)!;
+    courseRef.isFeedbackSessionsLoading = true;
+    courseRef.hasFeedbackSessionsLoadingFailed = false;
+    this.feedbackSessionsService.getFeedbackSessionsForStudent('student', courseId)
+      .subscribe((fss: FeedbackSessions) => {
+        const sortedFss: FeedbackSession[] = this.sortFeedbackSessions(fss);
+        const studentSessions: StudentSession[] = courseRef.feedbackSessions;
 
-            const studentSessions: StudentSession[] = [];
-            this.feedbackSessionsService.hasStudentResponseForAllFeedbackSessionsInCourse(course.courseId)
-              .pipe(finalize(() => {
-                this.isCoursesLoading = false;
-              }))
-              .subscribe((hasRes: HasResponses) => {
-                if (!hasRes.hasResponsesBySession) {
-                  this.statusMessageService.showErrorToast(this.allStudentFeedbackSessionsNotReturned);
-                  this.hasCoursesLoadingFailed = true;
-                  return;
-                }
+        this.feedbackSessionsService.hasStudentResponseForAllFeedbackSessionsInCourse(courseId)
+          .pipe(finalize(() => { courseRef.isFeedbackSessionsLoading = false; }))
+          .subscribe((hasRes: HasResponses) => {
+            if (!hasRes.hasResponsesBySession) {
+              this.statusMessageService.showErrorToast(this.allStudentFeedbackSessionsNotReturned);
+              this.hasCoursesLoadingFailed = true;
```

Comment:
Should this be `hasFeedbackSessionsLoadingFailed` instead? At this point courses should already be loaded. There's one more on line 150 that should probably be changed as well.

Reply:
Thanks for pointing that out. I have changed them in the next commit.